tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,305 @@
+
+    american_prices = tf.where(
+      condition,
+      eu_prices + a2 * (s / s_crit) ** q2,
+      (s - x) * sign)
+
+    return american_prices, converged, failed
+
+
+def _option_price_formula_variables(sigma,
+                                    x,
+                                    t,
+                                    r,
+                                    b,
+                                    sign,
+                                    is_call_options,
+                                    dtype,
+                                    max_iterations=20,
+                                    tolerance=1e-8):
+  """ Calculates variables q, a and the critical spot price as per reference [1]
+  for the american option price formula.
+  These are values needed to calculate the approximate american option price.
+  For spot prices above and below the critical spot price, the american option
+  price formula behaves differently.
+
+  The naming convention will align variables with the variables named in
+  reference [1], but made lower case.
+
+  ## References:
+  [1] Baron-Adesi, Whaley, Efficient Analytic Approximation of American Option
+    Values, The Journal of Finance, Vol XLII, No. 2, June 1987;
+    https://deriscope.com/docs/Barone_Adesi_Whaley_1987.pdf"""
+
+  m = 2 * r / sigma ** 2
+  n = 2 * b / sigma ** 2
+  k = 1 - tf.exp(- r * t)
+  q_ = _calc_q(n, m, sign, k)
+
+  value = lambda s_crit: vanilla_prices.option_price(
+    volatilities=sigma, strikes=x, expiries=t, spots=s_crit, discount_rates=r,
+    cost_of_carries=b, is_call_options=is_call_options, dtype=dtype) + sign * (
+      1 - tf.exp((b - r) * t) * _ncdf(sign * _calc_d1(
+    s_crit, x, sigma, b, t))) * s_crit / q_ - sign * (s_crit - x)
+
+  value_and_gradient = lambda price: tff.math.value_and_gradient(value, price)
+
+  # Calculate seed value for critical spot price
+  q_inf = _calc_q(n, m, sign)
+  s_inf = x / (1 - 1 / q_inf)
+  h = -(sign * b * t + 2 * sigma * tf.sqrt(t)) * sign * (x / (s_inf - x))
+  if is_call_options is None:
+    s_seed = x + (s_inf - x) * (1 - tf.exp(h))
+  else:
+    s_seed = tf.where(
+      is_call_options,
+      x + (s_inf - x) * (1 - tf.exp(h)),
+      s_inf + (x - s_inf) * tf.exp(h))
+
+  s_crit, converged, failed = implied_vol_newton.newton_root_finder(
+    value_and_grad_func=value_and_gradient, initial_values=s_seed,
+    max_iterations=max_iterations, tolerance=tolerance, dtype=dtype)
+
+  a_ = (sign * s_crit / q_) * (1 - tf.exp((b - r) * t) * _ncdf(
+    sign * _calc_d1(s_crit, x, sigma, b, t)))
+
+  return q_, a_, s_crit, converged, failed
+
+
+def _calc_d1(s, x, sigma, b, t):
+  return (tf.math.log(s / x) + (b + sigma ** 2 / 2) * t) / (
+      sigma * tf.math.sqrt(t))
+
+
+def _calc_q(n, m, sign, k=1):
+  return ((1 - n) + sign * tf.sqrt((n - 1) ** 2 + 4 * m / k)) / 2
+
+
+def _ncdf(x):
+  return (tf.math.erf(x / _SQRT_2) + 1) / 2
+
+
+def _npdf(x):
+  return tf.exp(-0.5 * x ** 2) / tf.math.sqrt(2 * np.pi)
```

Comment:
tf.math.sqrt(2 * np.pi) is always of dtype tf.float64. Could you please define a global constant like _SQRT_2?